src/transformers/modeling_utils.py

```diff
@@ -4911,64 +4996,66 @@ def _load_pretrained_model(
+        # Use multiprocessing Pool for parallel execution, off by default
+        if json.loads(os.environ.get("HF_ENABLE_PARALLEL_LOADING", "false")) and not is_deepspeed_zero3_enabled():
+            num_workers = json.loads(os.environ.get("HF_PARALLEL_LOADING_WORKERS", "8"))
 
-            # If shard_file is "", we use the existing state_dict instead of loading it
-            if shard_file != "":
-                state_dict = load_state_dict(
-                    shard_file, is_quantized=is_quantized, map_location=map_location, weights_only=weights_only
-                )
+            # Do not spawn anymore workers than you need
+            num_workers = min(len(args_list), num_workers)
 
-            # Fix the key names
-            state_dict = {key_renaming_mapping[k]: v for k, v in state_dict.items() if k in key_renaming_mapping}
+            logger.info(f"Loading model weights in parallel with {num_workers} workers...")
 
-            if is_deepspeed_zero3_enabled() and not is_quantized:
-                error_msgs += _load_state_dict_into_zero3_model(model_to_load, state_dict)
-            # Skip it with fsdp on ranks other than 0
-            elif not (is_fsdp_enabled() and not is_local_dist_rank_0() and not is_quantized):
-                disk_offload_index, cpu_offload_index = _load_state_dict_into_meta_model(
-                    model_to_load,
-                    state_dict,
-                    shard_file,
-                    expected_keys,
-                    reverse_key_renaming_mapping,
-                    device_map=device_map,
-                    disk_offload_folder=disk_offload_folder,
-                    disk_offload_index=disk_offload_index,
-                    cpu_offload_folder=cpu_offload_folder,
-                    cpu_offload_index=cpu_offload_index,
-                    hf_quantizer=hf_quantizer,
-                    is_safetensors=is_offloaded_safetensors,
-                    keep_in_fp32_regex=keep_in_fp32_regex,
-                    unexpected_keys=unexpected_keys,
-                    device_mesh=device_mesh,
-                )
+            # with multiprocessing.Pool(processes=num_workers) as pool:
+            with ThreadPoolExecutor(max_workers=num_workers) as executor:
+                with logging.tqdm(total=len(args_list), desc="Loading checkpoint shards") as pbar:
+                    futures = [executor.submit(load_shard_file, arg) for arg in args_list]
+                    for future in as_completed(futures):
+                        result = future.result()
+                        (
+                            _error_msgs,
+                            disk_offload_index,
+                            cpu_offload_index,
+                        ) = result
 
-            # force memory release if loading multiple shards, to avoid having 2 state dicts in memory in next loop
-            del state_dict
+                        error_msgs += _error_msgs
+
+                        pbar.update(1)
```

Comment:
can we put that in an external function? 
Just to not make from_pretrained bigger than it already is!